### ISSUE TITLE
[no-unnecessary-condition] `Partial<Record<L, T>>[L]` should be considered as maybe `falsy`

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3755]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": [2]
  }
}
```

```TS
function _getValue<L extends string, T>(
  source: Partial<Record<L, T>> | null,
  locale: L,
  loose?: boolean
) {
  if (!source) {
    return;
  }
  let value = source[locale];
  if (value == null && loose) {
  }
}
```

[ts playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAfQOYFMoDUCGAbEdAHgBlF0APKdMAEwGdF6oAnGMVAGkQBUA+ABQBYAFCImcECwjoAXIgAKOFrHxEASughwWtUt359EAH0RgQePJ1Hi8cCPjmIS1sYjtx66APzyARnBweOg4YKIAlIgA3jaIMMCIAgCE9JLS6JExbuIsmFJgANyxAL6xwVCIAG74hIgAvBJSMgDadg7BALpFbvGJ1QTo9Q3mlogAZGPugV6ZJaKlIkA)

![image](https://user-images.githubusercontent.com/8336744/129986240-61e4b877-a33a-473e-bd0a-22135747c92b.png)

```jsonc
{
  "extends": "@1stg/tsconfig/tsconfig",
  "compilerOptions": {
    "baseUrl": ".",
    "jsx": "react-jsx",
    "lib": ["DOM", "ESNext"],
    "paths": {
      "*": ["src/*"]
    }
  }
}
```

**Expected Result**

No warning/error

**Actual Result**

Error

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.2` |
| `@typescript-eslint/parser`        | `4.29.2` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.22.5` |

**Comment by bradzacher** [comment]
The type of `value` is reported by TS exactly as you see in the screenshot.
TS reports it to us as a unique type - `Partial<Record<L, T>>[L]`.

It does not report it to us as `T | undefined` - it just chooses not to resolve the complex generic type.
I had a quick look and there doesn't seem to be an API available which supports resolving this.

TS itself doesn't even handle this type well - it doesn't even narrow:
```ts
function _getValue<L extends string, T>(
  source: Partial<Record<L, T>>,
  locale: L
) {
  if (!source) {
    return;
  }
  let value = source[locale];
  if (value === undefined) {
     // typeof value === Partial<Record<L, T>>[L] ((id: 95))
     value;
  } else {
     // typeof value === Partial<Record<L, T>>[L] ((id: 95))
     value;
  }
}
```
[repl](https://ts-ast-viewer.com/#code/GYVwdgxgLglg9mABAfQOYFMoDUCGAbEdAHgBlF0APKdMAEwGdF6oAnGMVAGkQBUA+ABQAoREzggWEdAC5EABRwtY+IgCV0EOC1qlu-PpxGI8cCPhmISQgJSIA3kZjBEAgIT1xk9LYejRLTAkwAG4jAF8jPExEADd8QkQAXjEJKQBtEzMogF1Q0ScXOIJ0JMTk8Fp0YHZ0Wh8jPwB6RsQoAE8AB3Q4ZyKEsuSFJRgVdU1tXV4+PjSSbJcBGFpZAE4AVmtrBtE+9DzEMPI8ehLfP0Rm1s7u3viSgflFZTw1DS0dEj1p2fmBReXEOtNttYnd9hEwkA)

From what I can tell - nothing we can do about this, unfortunately.
